File: frontend/src/lib/edits.ts
export interface EditOperation {
  id: string;
  asset_id: string;
  operation_type: string;
  brush_type: string;
  strength: number;
  radius: number;
  position: number[];
  status: "queued" | "processing" | "completed" | "failed";
  provider: string;
  message: string;
  created_at: string;
  updated_at: string;
}

export const SCULPT_TOOLS = [
  { id: "clay",    label: "Clay",    description: "Build up surface volume" },
  { id: "smooth",  label: "Smooth",  description: "Relax and soften surface" },
  { id: "inflate", label: "Inflate", description: "Push surface outward" },
  { id: "pinch",   label: "Pinch",   description: "Pull surface toward center" },
  { id: "move",    label: "Move",    description: "Translate a surface region" },
  { id: "mirror",  label: "Mirror",  description: "Symmetric editing across axis" },
] as const;

export const FALLOFF_TYPES = ["sphere", "gaussian", "flat"] as const;

export async function createEditOperation(params: {
  asset_id: string;
  operation_type: string;
  brush_type: string;
  strength: number;
  radius: number;
  position?: number[];
}): Promise<EditOperation> {
  const res = await fetch("/api/edits/create", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(params),
  });
  if (!res.ok) {
    const err = await res.json().catch(() => ({ detail: "Create failed" }));
    throw new Error(err.detail ?? "Failed to create edit operation");
  }
  return res.json();
}

export async function getEditOperation(operationId: string): Promise<EditOperation> {
  const res = await fetch(`/api/edits/${operationId}`);
  if (!res.ok) throw new Error("Edit operation not found");
  return res.json();
}

export async function listEditOperations(assetId?: string): Promise<EditOperation[]> {
  const url = assetId ? `/api/edits?asset_id=${assetId}` : "/api/edits";
  const res = await fetch(url);
  if (!res.ok) throw new Error("Failed to list edit operations");
  return res.json();
}

File: backend/app/routes/edits.py
from __future__ import annotations

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from app.services.edit_service import (
    create_edit_operation,
    get_edit_operation,
    list_edit_operations,
)

router = APIRouter(prefix="/api/edits", tags=["edits"])


class CreateEditRequest(BaseModel):
    asset_id: str
    operation_type: str = "sculpt"
    brush_type: str = "clay"
    strength: float = 0.5
    radius: float = 20.0
    position: list[float] = []


@router.post("/create", status_code=201)
async def create_edit(req: CreateEditRequest):
    try:
        op = create_edit_operation(
            asset_id=req.asset_id,
            operation_type=req.operation_type,
            brush_type=req.brush_type,
            strength=req.strength,
            radius=req.radius,
            position=req.position,
        )
        return op
    except ValueError as e:
        raise HTTPException(status_code=422, detail=str(e))


@router.get("")
async def list_edits(asset_id: str | None = None):
    return list_edit_operations(asset_id=asset_id)


@router.get("/{operation_id}")
async def get_edit(operation_id: str):
    op = get_edit_operation(operation_id)
    if not op:
        raise HTTPException(status_code=404, detail="Edit operation not found")
    return op
